065-topic-replication-factor-change.md

```diff
@@ -0,0 +1,298 @@
+The `replicasChange` can be in a pending or ongoing `state`, and it can contain a `message` in case of failure.
+Other fields are `sessionId`, which maps to Cruise Control's `User-Task-ID`, and `targetReplicas`, which reflects the target replicas value (this may be different from .spec.replicas when the state is ongoing).
+
+- **Pending**: Not in Cruise Control's task queue (not yet sent or request error).
+  Cruise Control's task states: None.
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:50.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 2
+    replicasChange:
+      state: pending
+      targetReplicas: 2
+    topicName: my-topic
+  ```
+
+- **Ongoing**: In Cruise Control's task queue, but execution not started, or not completed. 
+  Cruise Control's task states: Active, InExecution.
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:53.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 3
+    replicasChange:
+      sessionId: 1aa418ca-53ed-4b93-b0a4-58413c4fc0cb
+      state: ongoing
+      targetReplicas: 2
+    topicName: my-topic
+  ```
+
+- **Completed**: Cruise Control's task execution completed (target replication factor reconciled).
+  Cruise Control's task states: Completed.
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:58.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 4
+    topicName: my-topic
+  ```
+
+- **Failed**: Cruise Control's request failed, the change remains pending, and it is periodically retried.
+  Cruise Control's task states: CompletedWithError.
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:58.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 4
+    replicasChange:
+      message: Change request failed, Cluster model not ready
+      state: pending
+      targetReplicas: 2
+    topicName: my-topic
+  ```
+
+### Reconciliation logic
+
+At runtime, the Topic Operator watches and periodically reconciles the configuration of all managed and unpaused KafkaTopic resources.
+Among the other configurations, the reconciliation logic detects `.spec.replicas` changes in KafkaTopic resources by comparing with the topic replication factor value in Kafka.
+
+A ReplicasChangeClient will be used to isolate the logic required to call the Cruise Control REST API from the rest of the reconciliation logic.
+The ReplicasChangeClient will support sending multiple replicas changes with a single request, and checking multiple tasks with a single request.
+
+On every reconciliation loop, when Cruise Control integration will be enabled and replicas changes are detected, the BatchingTopicController will use the ReplicasChangeClient for the following operations:
+
+1. Request pending replicas changes (uses `topic_configuration` endpoint, and returns topics with pending and ongoing changes).
```

Comment:
Somewhere you should describe the error case where we've filled up CC's internal queue. AFAICS what you're describing will work fine (assuming CC returns some error response in this situation): We'll eventually retry and work through the backlog of `KafkaTopics`.

Reply:
It is like you say. I have the following row, let me know if it's enough.

> Another temporary example is the `RuntimeException` when Cruise Control's task queue is full.